lib/kvdb/ikvdb.c

```diff
@@ -104,6 +106,7 @@ thread_local char tls_vbuf[256 * 1024] HSE_ALIGNED(PAGE_SIZE);
 const size_t      tls_vbufsz = sizeof(tls_vbuf);
 
 static atomic_int kvdb_alias;
+static bool kvdb_opened = false;
```

Comment:
Can two threads execute `hse_kvdb_open()` at the same time?

Reply:
wouldn't be hard to protect against it.

Reply:
Would an `atomic_bool` suffice? Or should I go for a mutex?

Reply:
An atomic_bool would suffice...

Reply:
Seems to me like a lot of the instantiation of the global caches and structures would happen during ikvdb_open(), so they should be protected by a mutex. But that is just a guess without digging into the code.

Reply:
Which caches?  All the global caches should be managed under hse_init() / hse_fini()...

Reply:
They used to all have reference counts so that they could be created at first kvdb-open and then simply incremented for additional kvdb-opens, but that caused a lot of thrashing if you open/closed a kvdb multiple times (plus Alex didn't like them IIRC).  Anyway, I think things are simpler under hse init/fini...

Reply:
I am just throwing out *potential* reasons for a mutex over an atomic_bool

Reply:
I think we should just fix the single-threaded issue once and for all and use a semphore (sem_init(3)) to ensure single-threaded access to all single-threaded interfaces (hse_init, hse_fini, hse_kvdb_open, hse_kvdb_close, ...)
When you hold have the semaphore then you can safely manage other variables (hse_initialized, kvdb_open, ...) and they need not be atomic...
Then we don't every have to debug some customer issue where they did in fact have two or more threads in libhse violating our constraints...